dmriprep/utils/grouping.py

```diff
@@ -0,0 +1,711 @@
+    >>> epi_fmap2 = {"epi": "/data/sub-1/fmap/sub-1_dir-AP_run-2_epi.nii.gz", "suffix": "epi"}
+    >>> epi_fmap3 = {"epi": "/data/sub-1/fmap/sub-1_dir-PA_epi.nii.gz", "suffix": "epi"}
+    >>>
+    >>> phasediff_fmap = {"phasediff": "/data/sub-1/fmap/sub-1_phasediff.nii.gz",
+    ...                   "suffix": "phasediff"}
+    >>> phases_fmap = {"phase1": "/data/sub-1/fmap/sub-1_phase1.nii.gz",
+    ...                "suffix": "phase1"}
+    >>>
+    >>> dwi_fmap1 = {"dwi": "/data/sub-1/dwi/sub-1_dir-AP_dwi.nii.gz", "suffix": "dwi"}
+    >>> dwi_fmap2 = {'suffix': 'dwi',
+    ...   'dwi': ['/data/sub-1/dwi/sub-1_dir-AP_run-1_dwi.nii.gz',
+    ...    '/data/sub-1/dwi/sub-1_dir-AP_run-2_dwi.nii.gz',
+    ...    '/data/sub-1/dwi/sub-1_dir-AP_run-3_dwi.nii.gz']}
+
+    When there are no fieldmaps in ``fmaps/``, but a reverse PE DWI series
+    >>> get_highest_priority_fieldmap([dwi_fmap1])
+    {'dwi': '/data/sub-1/dwi/sub-1_dir-AP_dwi.nii.gz', 'suffix': 'dwi'}
+
+    There is both an epi fieldmap and a phase1/phase2 GRE fieldmap
+    >>> get_highest_priority_fieldmap([epi_fmap1, phases_fmap])
+    {'suffix': 'epi', 'epi': ['/data/sub-1/fmap/sub-1_dir-AP_run-1_epi.nii.gz']}
+
+    Multiple EPI fieldmaps
+    >>> get_highest_priority_fieldmap(
+    ...     [epi_fmap1, epi_fmap2, epi_fmap3]) # doctest: +NORMALIZE_WHITESPACE
+    {'suffix': 'epi',
+     'epi': ['/data/sub-1/fmap/sub-1_dir-AP_run-1_epi.nii.gz',
+             '/data/sub-1/fmap/sub-1_dir-AP_run-2_epi.nii.gz',
+             '/data/sub-1/fmap/sub-1_dir-PA_epi.nii.gz']}
+
+    An EPI fieldmap from ``fmap/`` should be chosen over a reverse PE DWI series
+    >>> get_highest_priority_fieldmap([epi_fmap1, dwi_fmap2])
+    {'suffix': 'epi', 'epi': ['/data/sub-1/fmap/sub-1_dir-AP_run-1_epi.nii.gz']}
+
+    """
+    # Find fieldmaps
+    default_priority = max(FMAP_PRIORITY.values()) + 1
+    priority = default_priority
+    selected_fmap_info = {'suffix': None}
+
+    # collapse multiple EPI fieldmaps into one entry
+    epi_fmaps = sorted([fmap_info["epi"] for fmap_info in fmap_infos
+                        if fmap_info.get('suffix') == 'epi'])
+    if epi_fmaps:
+        epi_info = {"suffix": "epi", "epi": epi_fmaps}
+        fmap_infos = [fmap_info for fmap_info in fmap_infos
+                      if fmap_info.get('suffix') != 'epi'] + [epi_info]
+
+    # Select the highest priority fieldmap
+    for fmap_info in fmap_infos:
+        if fmap_info.get('suffix') == 'phase':
+            fmap_info['suffix'] = 'phase1'
+        fmap_type = fmap_info.get('suffix')
+        if fmap_type not in FMAP_PRIORITY:
+            continue
+        this_priority = FMAP_PRIORITY[fmap_type]
+        if this_priority < priority:
+            priority = this_priority
+            selected_fmap_info = fmap_info
+
+    return selected_fmap_info
+
+
+def find_fieldmaps_from_other_dwis(dwi_files, dwi_file_metadatas):
+    """Find a list of files in the dwi/ directory that can be used for distortion correction.
```

Comment:
It would be good to restrict line length to <80 characters, so this doesn't wrap around when looking at it on GitHub's side-by-side diffs.